_posts/2017-06-21-Devoxx-pl.md

```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
+
+>The integration should be validated every day and when the build breaks, fix it!
+
+For unfinished work we can wait till it is ready before checking in.
+This exposes us to the risk of losing work when it's only on the developers computer.
+
+### Feature Branch
+
+An alternative would be to create a **feature branch**, which brings us back to the problem of merging branches.
+
+>Pain of merge = fn(size_of_merge, duration_since_last_merge)
+
+Merging branches can be a difficult task and might lead to commit race, to offload the effort to a colleague.
+
+### Trunk-based development
+
+A third option would be to 'check in anyway', called **trunk-based development**.
+Every commit integrates to the trunk and developers should integrate their local changes daily.
+Small changes and integrating often makes it easier to merge new code.
+
+New half-finished features can be hidden with **feature toggles**.
+These toggles can me managed using flags or configurations (eg, in Zookeeper, Consul, ...).
+>A flag should be set and evaluated in as few places as possible, preferably once each.
+>Flags should be removed when the new implementation is done.
+
+More info: [Trunk-based development](https://trunkbaseddevelopment.com/)
+
+Changes to existing functionality can be done by providing an abstraction above the existing functionality.
+The new functionality can then be developed for the abstraction and when it is done, change to the new implementation.
+**Branch-by abstraction** has the side-benefit that it can be used for **A/B and canary releasing**.
+
+The Continuous Delivery book tells us to treat every check in as a possible release candidate.
+Developers start with the assumption it is worthy, CI tool decides if it is.
+Deploy frequently with small changes, makes easier to rollback, lowering the chance of problems.
+
+### Git
+
+And then there was **Git**, developed by Linus Torvalds with the goal to merge a patch in less than three seconds.
+Branches where much more lightweight and every repository contained the full source history.
+
+In 2008 **[GitHub](https://github.com)** was founded and introduced pull requests.
+If you wanted to contribute to open source projects before pull requests you had to:
+* develop it locally, 
+* generate patch file 
+* mail it over to the project owners.
+This feature contributed to GitHub's success, three years later in 2011 they passed SourceForge and Google Code in popularity.
+
+Sam made the remark that pull requests used branches, which might bring problems. 
+On top of that **[GitFlow](https://www.atlassian.com/git/tutorials/comparing-workflows#gitflow-workflow)** was introduced.
+Because GitFlow introduces even more branches it is in controversy with fast deployment and small changes cycle.
+With tools like [split.io](https://www.split.io) and [launch darkly](https://launchdarkly.com) Git-Flow is not needed, if merged frequently.
+
+To concluded Sam asked himself if branches where evil.
```

Comment:
You probably meant: To conclude, Sam challenged the audience and asked if branches are evil.